src/ethereum/ethEngine.js

```diff
@@ -0,0 +1,947 @@
+/**
+ * Created by paul on 7/7/17.
+ */
+// @flow
+
+import type {
+  EdgeTransaction,
+  EdgeSpendInfo,
+  EdgeCurrencyEngineOptions,
+  EdgeWalletInfo
+} from 'edge-core-js'
+import { error } from 'edge-core-js'
+import {
+  validateObject,
+  normalizeAddress,
+  toHex,
+  bufToHex,
+  getEdgeInfoServer,
+  addHexPrefix
+} from '../common/utils.js'
+import {
+  EtherscanGetBlockHeight,
+  EtherscanGetTransactions,
+  EtherscanGetAccountBalance,
+  EtherscanGetAccountNonce,
+  EtherscanGetTokenTransactions,
+  EthGasStationSchema,
+  NetworkFeesSchema,
+  SuperEthGetUnconfirmedTransactions
+} from './ethSchema.js'
+import { bns } from 'biggystring'
+
+import {
+  type EtherscanTransaction,
+  type EthereumTxOtherParams,
+  type EthereumFee,
+  type EthereumFeesGasPrice,
+  type EthereumWalletOtherData
+} from './ethTypes.js'
+import { EthereumPlugin } from './ethPlugin.js'
+import { CurrencyEngine } from '../common/engine.js'
+import { currencyInfo } from './ethInfo.js'
+import { calcMiningFee } from './ethMiningFees.js'
+
+const abi = require('./export-fixes-bundle.js').ABI
+const ethWallet = require('./export-fixes-bundle.js').Wallet
+const EthereumTx = require('./export-fixes-bundle.js').Transaction
+
+const PRIMARY_CURRENCY = currencyInfo.currencyCode
+const ACCOUNT_POLL_MILLISECONDS = 20000
+const BLOCKCHAIN_POLL_MILLISECONDS = 20000
+const TRANSACTION_POLL_MILLISECONDS = 20000
+const UNCONFIRMED_TRANSACTION_POLL_MILLISECONDS = 3000
+const NETWORKFEES_POLL_MILLISECONDS = 60 * 10 * 1000 // 10 minutes
+const ADDRESS_QUERY_LOOKBACK_BLOCKS = 4 * 60 * 24 * 7 // ~ one week
+const NUM_TRANSACTIONS_TO_QUERY = 50
+
+type BroadcastResults = {
+  incrementNonce: boolean,
+  decrementNonce: boolean
+}
+
+export class EthereumEngine extends CurrencyEngine {
+  ethereumPlugin: EthereumPlugin
+  otherData: EthereumWalletOtherData
+
+  constructor (
+    currencyPlugin: EthereumPlugin,
+    io_: any,
+    walletInfo: EdgeWalletInfo,
+    opts: EdgeCurrencyEngineOptions
+  ) {
+    super(currencyPlugin, io_, walletInfo, opts)
+    if (typeof this.walletInfo.keys.ethereumKey !== 'string') {
+      if (walletInfo.keys.keys && walletInfo.keys.keys.ethereumKey) {
+        this.walletInfo.keys.ethereumKey = walletInfo.keys.keys.ethereumKey
+      }
+    }
+    this.currencyPlugin = currencyPlugin
+  }
+
+  async fetchGetEtherscan (server: string, cmd: string) {
+    let apiKey = ''
+    if (global.etherscanApiKey && global.etherscanApiKey.length > 5) {
+      apiKey = '&apikey=' + global.etherscanApiKey
+    }
+    const url = `${server}/api${cmd}${apiKey}`
+    return this.fetchGet(url)
+  }
+
+  async fetchGet (url: string) {
+    const response = await this.io.fetch(url, {
+      method: 'GET'
+    })
+    if (!response.ok) {
+      const cleanUrl = url.replace(global.etherscanApiKey, 'private')
```

Comment:
Do we need to sanitize the blockcypher API key here too?

Reply:
Good point. We were never doing that even in the original ethereum plugin.